iib/workers/tasks/build.py

```diff
@@ -337,6 +342,157 @@ def _get_resolved_image(pull_spec):
+
+    :param str from_index: index image to get database file from.
+    :param str base_dir: base directory to which the database file should be saved.
+    :return: path to the copied database file.
+    :rtype: str
+    :raises IIBError: if any podman command fails.
+    """
+    data = skopeo_inspect(f'docker://{from_index}')
+    try:
+        db_path = data['Labels']['operators.operatorframework.io.index.database.v1']
+    except KeyError:
+        raise IIBError('Index image doesn\'t have the label specifying its database location.')
+    _copy_files_from_image(from_index, db_path, base_dir)
+    local_path = os.path.join(base_dir, os.path.basename(db_path))
+    return local_path
+
+
+def _serve_index_registry(db_path):
+    """
+    Locally start OPM registry service, which can be communicated with using gRPC queries.
+
+    Due to IIB's paralellism, the service can run multiple times, which could lead to port
+    binding conflicts. Resolution of port conflicts is handled in this function as well.
+
+    :param str db_path: path to index database containing the registry data.
+    :return: tuple containing port number of the running service and the running Popen object.
+    :rtype: (int, Popen)
+    :raises IIBError: if all tried ports are in use, or the command failed for another reason.
+    """
+    conf = get_worker_config()
+    port = conf['iib_grpc_start_port']
+    for i in range(conf['iib_grpc_max_port_tries']):
+        cmd = ['opm', 'registry', 'serve', '-p', str(port), '-d', db_path]
+        registry_initialized = False
+        for j in range(conf['iib_grpc_max_tries']):
+            rpc_proc = subprocess.Popen(
+                cmd, stdout=subprocess.PIPE, stderr=subprocess.PIPE, universal_newlines=True
+            )
+            time.sleep(conf['iib_grpc_init_wait_time'])
+            ret = rpc_proc.poll()
+            # process hasn't terminated
+            if ret is None:
+                stdout = get_running_subprocess_output(rpc_proc.stdout)
+                if 'serving registry' in stdout:
+                    registry_initialized = True
+                    log.debug('Started the command "%s"', ' '.join(cmd))
+                    break
+                elif j == conf['iib_grpc_max_tries'] - 1:
+                    rpc_proc.kill()
+                    raise IIBError(
+                        'Index registry has not been initialized after '
+                        f'{conf.get("iib_grpc_max_tries")} tries'
+                    )
+                # not yet initialized, retry the loop
+                else:
+                    # sometimes, the registry service doesn't output any messages despite running
+                    # restarting it is an attempt to mitigate this issue
+                    rpc_proc.kill()
+                    continue
+            # process has terminated, something went wrong
+            else:
+                break
+        if registry_initialized:
```

Comment:
Could you please add a log message here for easy debugging?

Reply:
done